extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java

```diff
@@ -93,10 +113,20 @@ BsonDiscriminatorBuildItem collectBsonDiscriminators(CombinedIndexBuildItem inde
     }
 
     @BuildStep
-    CommandListenerBuildItem collectCommandListeners(CombinedIndexBuildItem indexBuildItem) {
+    CommandListenerBuildItem collectCommandListeners(CombinedIndexBuildItem indexBuildItem, Capabilities capabilities,
+            MongoClientBuildTimeConfig buildTimeConfig) {
+        Predicate<String> isTracingCommandListenerPredicate = name -> name
+                .equalsIgnoreCase("io.opentracing.contrib.mongo.common.TracingCommandListener");
         Collection<ClassInfo> commandListenerClasses = indexBuildItem.getIndex()
                 .getAllKnownImplementors(DotName.createSimple(CommandListener.class.getName()));
-        List<String> names = commandListenerClasses.stream().map(ci -> ci.name().toString()).collect(Collectors.toList());
+        List<String> names = commandListenerClasses.stream()
+                .map(ci -> ci.name().toString())
+                // filter the TracingCommandListener to avoid the non-default constructor class.
+                .filter(isTracingCommandListenerPredicate.negate())
+                .collect(Collectors.toList());
+        if (buildTimeConfig.tracingEnabled && capabilities.isPresent(Capability.OPENTRACING)) {
+            names.add(MongoTracingCommandListener.class.getName());
```

Comment:
You should index your `MongoTracingCommandListener` if tracing is enabled and the capability is present in a dedicated step. By this you will not have to remove the TracingCommandListene and add your own

Reply:
After removing the indexing, the filtering was no more necessary.

But I was not able (didn't know how) to extract the indexing of the `MongoTracingCommandListener` in a separate BuildStep. I did it but I got a cycle detection error when including the Jackson extension (between the capabilities and the combined index due to the JacksonProcessor.register). So finally I decide to keep the manually adding part but if the dedicated step is required I will need some hints about how to do it.